File: C++/N-Queens.cpp
class Solution {
public:
    void helper(vector<vector<string> >& res, vector<string>& cur, 
        vector<vector<int> >& vis, int pos, int n) {
        if (pos >= n) {
            res.push_back(cur);
            return;
        }
        for (int i = 0; i < n; i ++) {
            if (vis[0][i] == 0 && vis[1][n + i - pos] == 0 && vis[2][i + pos] == 0) {
                vis[0][i] = vis[1][n + i - pos] = vis[2][i + pos] = 1;
                cur[pos][i] = 'Q';
                helper(res, cur, vis, pos + 1, n);
                cur[pos][i] = '.';
                vis[0][i] = vis[1][n + i - pos] = vis[2][i + pos] = 0;
            }
        }
    }
    vector<vector<string> > solveNQueens(int n) {
        vector<vector<string> > res;
        vector<string> cur;
        string str = "";
        for (int i = 0; i < n; i ++) {
            str += '.';
        }
        for (int i = 0; i < n; i ++) {
            cur.push_back(str);
        }
        vector<vector<int> > vis(3, vector<int>(30, 0));
        helper(res, cur, vis, 0, n);
        return res;
    }
};

File: Python/N-Queens.py

class Solution:
    # @return a list of lists of string
    def generate(self, pos, n, seq, vis, res):
        if pos == n:
            cnt = []
            for i in range(n):
                st = ""
                for j in range(n):
                    if seq[i][j] == 0:
                        st += "."
                    else:
                        st += "Q"
                cnt.append(st)
            res.append(cnt)
            return
        for i in range(n):
            if not vis[0][i] and not vis[1][pos + i] and not vis[2][pos - i + n]:
                vis[0][i] = vis[1][pos + i] = vis[2][pos - i + n] = 1
                seq[pos][i] = 1
                self.generate(pos+1, n, seq, vis, res)
                seq[pos][i] = 0
                vis[0][i] = vis[1][pos + i] = vis[2][pos - i + n] = 0
        return
    def solveNQueens(self, n):
        seq = [[0 for col in range(n)] for row in range(n)]
        vis = [[0 for col in range(30)] for row in range(3)]
        res = []
        self.generate(0, n, seq, vis, res)
        return res